sql: test + enable ruby

madelynnblue opened the pull request:
The change in v3.go allows types of '0', which the spec says to treat
as unspecified.

The changes to server hardcode a postgres port of 15432. We can add options
to change it later, but it is currently the same logic as the cockroach port.

The acceptance tests spawn a docker container running ruby with the pg
gem installed.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3602)

<!-- Reviewable:end -->

mjibson commented:
The docker invocation needs some thought. It seems a bit weird to just export `newDockerClient()` in the `acceptance/cluster` package because it doesn't make sense for that function to live there officially. But maybe it's fine. Thoughts?

tschottdorf commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[acceptance/cluster/docker.go, line 313 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GKrld-WIa61-lTpyP:-K7GKrld-WIa61-lTpyQ:-1762998385) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/cluster/docker.go#L313)):*
mind adding this to the `cluster.Cluster` interface and implementing it for the other impl (`terrafarm.Farmer`)? The signature should be
`func (Cluster) PGAddr(i int) *net.TCPAddr`.

---

*[acceptance/ruby_test.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GLKaeUrmLyvj6DCFi:-K7GLKaeUrmLyvj6DCFj:560964556) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/ruby_test.go#L39)):*
`t.Skip("skipping since not running in Docker mode")`. But really this will already have spawned a remote cluster by the time you get here so better use the way the other local-only tests do it:

```go
	if *numLocal == 0 {
		t.Skip("skipping since not run against local cluster")
	}
	l := cluster.CreateLocal(*numLocal, *numStores, *logDir, stopper) // intentionally using local cluster
	l.Start()
	defer l.AssertAndStop(t)

```

maybe throw in `	checkRangeReplication(t, l, 20*time.Second)` too if you want to make sure replication has happened before you start.

Feel free to pull this functionality out so that more is shared. For example

```go
func MustStartLocal(t *testing.T) *cluster.LocalCluster{
  	if *numLocal == 0 {
		t.Skip("skipping since not run against local cluster")
	}
	l := cluster.CreateLocal(*numLocal, *numStores, *logDir, stopper) // intentionally using local cluster
	l.Start()
	return l
}
```

so that, for example, `TestRangeReplication` becomes

```go
func TestRangeReplication(t *testing.T) {
	l := MustStartLocal(t)
	defer l.AssertAndStop(t)

	checkRangeReplication(t, l, 20*time.Second)
}
```

---

*[acceptance/ruby_test.go, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GMxFm3QoCFxiA78ml:-K7GMxFnQpgWDFmJKqLF:-865337317) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/ruby_test.go#L44)):*
I think it's fine to export `NewDockerClient` from `cluster`. A lot of things are in the wrong place already. Only having looked at it (again) for a few minutes now, I'm thinking `package cluster->dockroach` (or something that sounds like docker but doesn't shadow a package name), and moving `cluster.Cluster` to `acceptance.Cluster`. May be missing something though.

---

*[acceptance/ruby_test.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GOJ20J1BjJdwmXuuR:-K7GOJ21AUUhwD9T7_22:-600405094) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/ruby_test.go#L70)):*
you can't just replace `bash` above by `ruby` and feed it the program below and get an exit status? Got a one-liner to play with?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 6 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[server/context.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Gk7TECweSYbgLz8om:-K7Gk7TF4emc15khcW_f:816836656) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/server/context.go#L40)):*
this should be an integer; when we wire it up to a flag we're gonna want the integer behaviour. also, let's go ahead and make this a `defaultPGAddr` because it should be possible to have this listen on a different interface than everything else.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

bdarnell commented on server/context.go:
```diff
@@ -37,6 +37,7 @@ import (
 // Context defaults.
 const (
 	defaultAddr               = ":26257"
+	defaultPGPort             = "15432"
```
We shouldn't add this without making it configurable via a flag, since a hard-coded value makes it impossible to run multiple servers on the same machine, or to run the tests while a local server is running.

bdarnell commented on acceptance/ruby_test.go:
```diff
@@ -0,0 +1,79 @@
+// Invoking ruby directly in docker, I can't figure out how to get its
+// exit status.
+const ruby = `
+ruby << 'EOF'
```
Isn't this throwing away ruby's return value? You need `set -e` or `exec ruby` for the exit status to be propagated to the shell.

mjibson commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[acceptance/cluster/docker.go, line 313 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GKrld-WIa61-lTpyP:-K7IbllBE04-pDSXPUDP:-179132742) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/cluster/docker.go#L313)):*
Does this still need to happen with the MustStartLocal func below?

---

*[acceptance/ruby_test.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GLKaeUrmLyvj6DCFi:-K7IcI0J35vYKArGLoic:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/ruby_test.go#L39)):*
Done.

---

*[acceptance/ruby_test.go, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GMxFm3QoCFxiA78ml:-K7IayrfFTVi5k1a-ZIn:359131887) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/ruby_test.go#L44)):*
Just exporting it for now. The refactor can wait for another PR.

---

*[acceptance/ruby_test.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GOJ20J1BjJdwmXuuR:-K7Ib2-wlbiUSRDjx-Mg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/ruby_test.go#L70)):*
Done.

---

*[acceptance/ruby_test.go, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7FkLaIWa_Jyf0tirIQ-r1-72:-K7Hz6zpqP8nZOr7Ic1C:-227388032) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/ruby_test.go#L72)):*
New code does this correctly.

---

*[server/context.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7FkLaHcR-qYH6v59pc-r1-40:-K7Ib81DD7E5hZKJD0JX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/server/context.go#L40)):*
Done.

---

*[server/context.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Gk7TECweSYbgLz8om:-K7Ib8aoF8SDKkwJbM5C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/server/context.go#L40)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/cluster/docker.go, line 313 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GKrld-WIa61-lTpyP:-K7IkU7aCJuITQwdOLD8:1970173295) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/cluster/docker.go#L313)):*
yes, both local and remote clusters will expose Postgres and it's better not to create incompatible interfaces in the first place. You can put a dummy impl (just panic) for the remote cluster if it's not trivial to you what the impl should be.

---

*[acceptance/main_test.go, line 134 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Il8Pig8sJLZwaFWiI:-K7Il8Pig8sJLZwaFWiJ:-1373445842) ([raw file](https://github.com/cockroachdb/cockroach/blob/08533aece5550a3faf35dca5859edd0d7d3ee665/acceptance/main_test.go#L134)):*
that comment should go.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 9 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/cluster/docker.go, line 313 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GKrld-WIa61-lTpyP:-K7IltkFWMxFOx8Z5f3g:-168193557) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/cluster/docker.go#L313)):*
My argument against that is we are already limiting this to the LocalCluster impl only.  What is the point of adding a method on the interface when the other implementations don't need it and are just a panic?

---

*[acceptance/main_test.go, line 134 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Il8Pig8sJLZwaFWiI:-K7ImQ9ZF00HIgAXJB75:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/08533aece5550a3faf35dca5859edd0d7d3ee665/acceptance/main_test.go#L134)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 9 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[acceptance/cluster/docker.go, line 313 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GKrld-WIa61-lTpyP:-K7IoZAo8xc_xdqHMJz7:-1438616670) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/cluster/docker.go#L313)):*
The point is that there will be another implementation and the offer to just `panic` for now is to make life easier for you in this PR. If it still doesn't make sense don't worry about it now. It's not going to be much work for the person eventually doing it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
LGTM

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-:-K7IpItDtmdYi63NY9y-:676834233)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 9 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[acceptance/cluster/docker.go, line 313 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GKrld-WIa61-lTpyP:-K7Ipdsdt52q6YJeSopY:-559208144) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/cluster/docker.go#L313)):*
Ok, I added it. Was simple enough.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[acceptance/cluster/docker.go, line 313 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7GKrld-WIa61-lTpyP:-K7J7GKS7MTmdo0pZHoV:-218857341) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef76e32e56f5fd368a77435988f02620bb5ce0a/acceptance/cluster/docker.go#L313)):*
cool, thanks.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 9 files at r2, 1 of 1 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/pgwire/server.go, line 76 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7J8OxlNXWwlPDBTG_h:-K7J8Oxm4a_Vm3eGD5WI:-378341873) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f0036783ef4ac3a75c5150e8d1199598d2961c/sql/pgwire/server.go#L76)):*
intended?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/pgwire/server.go, line 76 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7J8OxlNXWwlPDBTG_h:-K7JQGFSxZlE5oKyeGSZ:2009107706) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f0036783ef4ac3a75c5150e8d1199598d2961c/sql/pgwire/server.go#L76)):*
Yup. We print the same message for the cockroach server, so why not here too?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 9 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[acceptance/ruby_test.go, line 73 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Nn2lllk6RTIDcUqCh:-K7Nn2lllk6RTIDcUqCi:-698204914) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f0036783ef4ac3a75c5150e8d1199598d2961c/acceptance/ruby_test.go#L73)):*
something something error handling

---

*[acceptance/ruby_test.go, line 75 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Nmzc3-nxIA7x5JkYw:-K7Nmzc4tx2pSmVbMG97:1226871969) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f0036783ef4ac3a75c5150e8d1199598d2961c/acceptance/ruby_test.go#L75)):*
don't you want this in the error message below?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: test + enable ruby, add pgaddr flag

The change in v3.go allows types of '0', which the spec says to treat
as unspecified.

The changes to server and cli listen by default on :15432 for postgres
traffic. Note that unlike the cockroach port, this binds (by default)
on all interfaces since the host is omitted.

The acceptance tests spawn a docker container running ruby with the pg
gem installed.

mjibson commented:
Review status: 11 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[acceptance/ruby_test.go, line 73 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Nn2lllk6RTIDcUqCh:-K7NuPLfwWQOV8BGvwQ6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f0036783ef4ac3a75c5150e8d1199598d2961c/acceptance/ruby_test.go#L73)):*
Done.

---

*[acceptance/ruby_test.go, line 75 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602#-K7Nmzc3-nxIA7x5JkYw:-K7NshyIh1ba5ffmB3Jx:-908617590) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8f0036783ef4ac3a75c5150e8d1199598d2961c/acceptance/ruby_test.go#L75)):*
The docker logs are many lines and it makes more sense to me to have them be a separate call. It ends up being the same either way.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3602)*
<!-- Sent from Reviewable.io -->
